src/Ranking/GameStatsGenerator.php

```diff
@@ -0,0 +1,609 @@
+<?php
+
+declare(strict_types=1);
+
+namespace EtoA\Ranking;
+
+use EtoA\Building\BuildingRepository;
+use EtoA\Defense\DefenseRepository;
+use EtoA\Race\RaceDataRepository;
+use EtoA\Ship\ShipRepository;
+use EtoA\Technology\TechnologyRepository;
+use EtoA\Universe\Planet\PlanetRepository;
+use EtoA\Universe\Planet\PlanetTypeRepository;
+use EtoA\Universe\Star\SolarTypeRepository;
+use EtoA\User\UserPropertiesRepository;
+use Exception;
+
+class GameStatsGenerator
+{
+    private const GAME_STATS_FILE = CACHE_ROOT . "/out/gamestats.html";
+
+    private PlanetTypeRepository $planetTypeRepository;
+    private SolarTypeRepository $solarTypeRepository;
+    private RaceDataRepository $raceDataRepository;
+    private PlanetRepository $planetRepository;
+    private BuildingRepository $buildingRepository;
+    private TechnologyRepository $technologyRepository;
+    private ShipRepository $shipRepository;
+    private DefenseRepository $defenseRepository;
+    private UserPropertiesRepository $userPropertiesRepository;
+
+    public function __construct(
+        PlanetTypeRepository $planetTypeRepository,
+        SolarTypeRepository $solarTypeRepository,
+        RaceDataRepository $raceDataRepository,
+        PlanetRepository $planetRepository,
+        BuildingRepository $buildingRepository,
+        TechnologyRepository $technologyRepository,
+        ShipRepository $shipRepository,
+        DefenseRepository $defenseRepository,
+        UserPropertiesRepository $userPropertiesRepository
+    ) {
+        $this->planetTypeRepository = $planetTypeRepository;
+        $this->solarTypeRepository = $solarTypeRepository;
+        $this->raceDataRepository = $raceDataRepository;
+        $this->planetRepository = $planetRepository;
+        $this->buildingRepository = $buildingRepository;
+        $this->technologyRepository = $technologyRepository;
+        $this->shipRepository = $shipRepository;
+        $this->defenseRepository = $defenseRepository;
+        $this->userPropertiesRepository = $userPropertiesRepository;
+    }
+
+    public function readCached(): ?string
+    {
+        return is_file(self::GAME_STATS_FILE) ? file_get_contents(self::GAME_STATS_FILE) : null;
+    }
+
+    public function generateAndSave(): void
+    {
+        $file = self::GAME_STATS_FILE;
+
+        $dir = dirname($file);
+        if (!is_dir($dir)) {
+            mkdir($dir, 0777, true);
+        }
+
+        if (file_put_contents($file, $this->generate()) === false) {
+            throw new Exception("Error! Could not write game stats file $file!");
+        }
+    }
+
+    public function generate(): string
```

Comment:
I think for this method here it would be great to have a smoke test that just asserts that the method returns a non empty string. In case we ever update any of the queries we know that it at least won't error after we deploy

Reply:
done 5ee23ef